// react-native-collapsible-tab-view/lib/commonjs/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isRTL = exports.ONE_FRAME_MS = exports.IS_IOS = exports.AnimatedSectionList = exports.AnimatedFlatList = void 0;
exports.scrollToImpl = scrollToImpl;
var _reactNative = require("react-native");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/** The time one frame takes at 60 fps (16 ms) */
const ONE_FRAME_MS = exports.ONE_FRAME_MS = 16;

/** check if app is in RTL mode or not */
const {
  isRTL
} = _reactNative.I18nManager;
exports.isRTL = isRTL;
const IS_IOS = exports.IS_IOS = _reactNative.Platform.OS === 'ios';
const AnimatedFlatList = exports.AnimatedFlatList = _reactNativeReanimated.default.createAnimatedComponent(_reactNative.FlatList);
const AnimatedSectionList = exports.AnimatedSectionList = _reactNativeReanimated.default.createAnimatedComponent(_reactNative.SectionList);
function scrollToImpl(ref, x, y, animated) {
  'worklet';

  if (!ref) return;
  // ensure we don't scroll on NaN
  if (!Number.isFinite(x) || !Number.isFinite(y)) return;
  (0, _reactNativeReanimated.scrollTo)(ref, x, y, animated);
}
